src/NumericalAlgorithms/LinearOperators/PowerMonitors.cpp

```diff
@@ -212,6 +236,77 @@ double convergence_rate(const DataVector& power_monitor,
   return -num / denom;
 }
 
+double number_of_pile_up_modes(const DataVector& power_monitor,
```

Comment:
This seems to do a decent amount of math that heavily overlaps with the convergence rate computation. It seems at least all the logs, would it make sense to combine the functions, or maybe offer a wrapper that does the shared math, for these implementations?

Probably a future issue, but I worry a bit about the number of memory allocations being done here because this is call per dim (and if I'm reading right per tensor component)?? Between this and the `convergence_rate`, that's a lot of memory allocations being done. Sure, this is only once every 50 time steps or something, but if this code alone does as many memory allocations as those 50 time steps, that's not great...